### ISSUE TITLE
YOLOv10 or YOLO11 Pruning, Masking and Fine-Tuning

### ISSUE THREAD

**Issue by Thaising-Taing** [ultralytics/ultralytics#19476]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I have a question. After pruning, I wanna fine-tuning with the trainer and apply masked to not let the weight update to update my prune weight. I applied this code, but it seems the mAP is too poor. Is there something wrong with my applied code? 

Looking forward your suggestions! 

    def create_pruning_mask(self):
        """Recreate the pruning mask based on the zero weights in the model."""
        masks = {}
        for name, module in self.model.named_modules():
            if isinstance(module, torch.nn.Conv2d):
                # Create a mask where weights are non-zero
                mask = (torch.abs(module.weight) > 1e-6).float()
                masks[name] = mask
        return masks

    def apply_pruned_mask(self):
        """Apply the masks to block gradients of pruned weights during training."""
        for name, module in self.model.named_modules():
            if isinstance(module, torch.nn.Conv2d) and name in self.masks:
                if module.weight.grad is not None:
                    mask = self.masks[name].to(module.weight.device)
                    if mask.size() != module.weight.grad.size():
                        raise ValueError(f"Mask and gradient size mismatch in layer {name}: "
                                        f"mask {mask.size()}, grad {module.weight.grad.size()}")
                    module.weight.grad *= mask

    def optimizer_step(self):
        """Perform a single step of the training optimizer with gradient clipping and EMA update."""
        self.scaler.unscale_(self.optimizer)  # unscale gradients
        self.apply_pruned_mask()    # Apply pruning mask
        torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)  # clip gradients
        self.scaler.step(self.optimizer)
        self.scaler.update()
        self.optimizer.zero_grad()
        if self.ema:
            self.ema.update(self.model)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Thaising-Taing, thank you for bringing this up and for your detailed explanation! 🚀

We suggest checking out the [Ultralytics Docs](https://docs.ultralytics.com/) for further guidance. You might find valuable insights under sections like [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) and [Custom Usage](https://docs.ultralytics.com/usage/python/). Additionally, your query about masking and pruning might also align with more advanced customization topics, and the [Discussions](https://github.com/orgs/ultralytics/discussions) might have complementary insights from the community.

To help us assist you better, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) of the issue? This could include the specific training steps you're using, dataset details, or any relevant logs. Your code snippet is helpful but testing it in a controlled, condensed form could help pinpoint what might be causing the poor mAP results 🐛.

In the meantime, make sure your environment is up to date to rule out version-related issues. Upgrade to the latest `ultralytics` package and dependencies using:

```bash
pip install -U ultralytics
```
Ensure your Python environment is **Python>=3.8** with compatible [PyTorch](https://pytorch.org/) versions as described in the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

Also, if you’re exploring experimental ideas like these, you may find it useful to test directly in any of these supported, GPU-enabled environments:
- **Colab Notebook**: <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Gradient Notebook**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a>
- **Kaggle Notebook**: <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Lastly, we encourage you to engage with the Ultralytics community for faster feedback and diverse perspectives:
- Real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧
- Share ideas or learn from others on [Discourse](https://community.ultralytics.com/)
- Explore discussions on [Subreddit](https://reddit.com/r/Ultralytics)

This is an automated response to get you started, but don't worry, an Ultralytics engineer will review this and follow up with additional guidance soon. Thanks for your patience and for being part of the Ultralytics community! 😊

**Comment by glenn-jocher** [comment]
For pruning and fine-tuning YOLO models, we recommend using Ultralytics' native pruning tools and following our Model Pruning and Fine-Tuning guides. Custom pruning implementations can interfere with YOLO's architecture-specific optimizations, potentially leading to performance drops. 

Key considerations:
1. Our YOLO models require structured pruning approaches that maintain channel alignment in concatenation-based architectures
2. Masked gradients should be applied before EMA updates and gradient clipping
3. Pruning thresholds and fine-tuning schedules need task-specific calibration

For best results, please follow our documented pruning workflow in the Ultralytics Model Pruning guide and YOLO Common Issues troubleshooting page. If you need commercial-grade pruning support, we offer Enterprise License holders prioritized technical assistance.

**Comment by Thaising-Taing** [comment]
Thank you for your guidances regarding to your native pruning tools. However, if I wanna fine-tuning the YOLOv10 or 11 to recover mAP, how many epochs do you recommend or we have to full-train the model? 

I would appreciate if you could share your thoughts. 

Thanks once again.

**Comment by glenn-jocher** [comment]
@Thaising-Taing for YOLOv10/YOLO11 fine-tuning after pruning, we recommend starting with 30-50% of the original training epochs (e.g., 100-150 epochs if originally trained for 300) while monitoring validation mAP. Our Model Pruning guide shows specific recovery schedules. For commercial implementations requiring guaranteed performance, Enterprise License holders receive optimized pruning configurations through prioritized support.

**Comment by Thaising-Taing** [comment]
Excuse me @glenn-jocher Could you mind point out where is the pruning guidelines that you refer to? I checked yesterday, i found this tutorial from your community https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity/

It seems this refer to YOLOv5, and I checked in YOLOv10 and 11, there has no this prune function. As well as, there has no tutorial to fine-tuning the model with masking the gradient. 

I would appreciate if you could share the tutorial how to prune and fine-tuning with masking. 

Thanks!

**Comment by glenn-jocher** [comment]
For YOLOv10/YOLO11 pruning implementation details, we recommend our YOLO Common Issues guide which covers structured pruning best practices. Native pruning API support is actively being developed - please watch GitHub Releases for updates. Enterprise License holders can contact us for early access to advanced pruning tools and dedicated support.
